File: Python/Obpysidian/generator/schedule.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import datetime
import location
import name
import schema.schema


class Schedule:

    def generate(self, date, latitude, longitude, radius, no_resources, no_activities, days, start, duration, language):
        ng = name.NameGenerator(language)
        lg = location.LocationGenerator()
        location_id = 0
        s = schema.schema.Schedule()

        # Resources
        for i in range(no_resources):
            location_ids = 'L%s' % location_id
            id = 'R%s' % (i + 1)
            (lat, lon) = lg.generate(latitude, longitude, radius)
            (first_name, last_name) = ng.generate()[:2]
            self.build_location(s, location_ids, lat, lon)
            self.build_resource(s, id, location_ids, first_name, last_name)
            self.build_shift(s, id, date, days, start, duration)
            location_id += 1

        # Activities
        for i in range(no_activities):
            location_ids = 'L%s' % location_id
            id = 'A%s' % (i + 1)
            (lat, lon) = lg.generate(latitude, longitude, radius)
            self.build_location(s, location_ids, lat, lon)
            self.build_activity(s, id, location_ids)
            location_id += 1

        return s

    def build_location(self, s, id, latitude, longitude):
        l = schema.schema.Location()
        l.id = id
        l.latitude = latitude
        l.longitude = longitude
        s.add_location(l)

    def build_resource(self, s, id, location_id, first_name, last_name):
        r = schema.schema.Resource()
        r.id = id
        r.location_id = location_id
        r.first_name = first_name
        r.last_name = last_name
        s.add_resource(r)

    def build_shift(self, s, resource_id, date, days, start, duration):
        for day in range(days):
            sh = schema.schema.Shift()
            sh.resource_id = resource_id
            sh.start = date + datetime.timedelta(days=day) + start
            sh.end = sh.start + duration
            sh.id = "%s-%s" % (resource_id, sh.start)
            s.add_shift(sh)

    def build_activity(self, s, id, location_id):
        a = schema.schema.Activity()
        a.id = id
        a.location_id = location_id
        s.add_activity(a)
        sa = schema.schema.ActivityStatus()
        sa.id = id
        sa.time = datetime.datetime.now()
        sa.duration = datetime.timedelta(minutes=30)
        s.add_activity_status(sa)

File: Go/Obsidian/generator/schedule.go
package generator

import (
	"../schema"
	"../shared/random"
	"./location"
	"./name"
	"fmt"
	"time"
)

func GenerateSchedule(latitude float64, longitude float64, radius float64, noResources int, noActivities int, language string) *schema.Schedule {
	s := schema.NewSchedule()
	r := random.GetRandom()
	ng := namegenerator.New(language, r)
	lid := 0
	dt := time.Now()

	// Resources
	for i := 0; i < noResources; i++ {
		lids := fmt.Sprintf("L%d", lid)
		id := fmt.Sprintf("R%d", i)

		// Location
		lat, lon := locationgenerator.Generate(r, latitude, longitude, radius)
		schema.NewLocation(s, lids, lat, lon)

		// Resource
		nng, first, last, _ := namegenerator.Generate(ng)
		ng = nng
		r := *schema.NewResource(nil, id)
		r.FirstName = first
		r.LastName = last
		r.LocationID = lids
		s.Resource[id] = r
		lid++
	}

	// Activities
	for i := 0; i < noActivities; i++ {
		lids := fmt.Sprintf("L%d", lid)
		id := fmt.Sprintf("A%d", i)

		// Location
		lat, lon := locationgenerator.Generate(r, latitude, longitude, radius)
		schema.NewLocation(s, lids, lat, lon)

		// Activity
		a := *schema.NewActivity(nil, id, dt, time.Duration(30*time.Minute))
		a.LocationID = lids
		s.Activity[id] = a
		lid++
	}
	return s
}
